src/com/cdd/bao/template/Vocabulary.java

```diff
@@ -461,8 +461,10 @@ else if (predicate.equals(externalURL))
 			else if (predicate.equals(altLabel1) || predicate.equals(altLabel2) || predicate.equals(altLabel3) || predicate.equals(altLabel4))
 			{
 				String[] list = uriToAlternateLabels.get(uri);
-				uriToAlternateLabels.put(uri, ArrayUtils.add(list, label));
-			}			
+				String primaryLabel = uriToLabel.get(uri); // get primaryLabel for this uri to check for redundancy against this putatuve altLabel
+				if (!ArrayUtils.contains(list, label) && !label.equals(primaryLabel)) // doesn't add label to the altLabels list for this uri if it already contains it as an altLabel or it's the same as a primaryLabel
+					uriToAlternateLabels.put(uri, ArrayUtils.add(list, label));
+			}
```

Comment:
Why do you think it's necessary to handle the primaryLabel as a special case. The code before didn't do it.

Reply:
before I was getting the URI using the label but I think that was backwards...
if (!(labelToURI.get(label).contains(label))) // ingores adding altLabels that are already a primaryLabel
 +					uriToAlternateLabels.put(uri, set); // add this label to the growing list of altLabels associated with this uri

Reply:
Only one of the two conditions applies: you should not do the "primaryLabel" part here, because you can't be sure of the order that it comes in, so that's broken. However: checking to make sure that "list" does not already contain "label" is correct.

The 2nd part - making sure alternate labels don't contain the primary label - should be done in the other code section that I indicated in the issue, i.e. the part where it writes it out to the vocab. This will mean that in the template editor, where it rederives the vocabulary, you'll see cases where an altLabel matches a label, and that's fine because it's relatively raw. But when it gets to the BAE via vocab.dump, it will not.